Feature/download fridge

Ylanv opened the pull request:
# Description

This feature ensures that ingredients stored in the user's fridge are automatically downloaded when the app is launched with an internet connection. This allows users to access their fridge's contents even when offline. Below is a video demonstrating this feature:

https://github.com/user-attachments/assets/384b08b1-55ed-42e9-a901-383b526a7bcc

---

# What has been changed?

## Fridge Item Local Storage
To enable offline access to fridge content, a new local database was implemented specifically for `FridgeItem`. This database stores details such as:
- **Barcode**
- **Quantity**
- **Expiration Date**

Here is a comprehensive schema detailing how a FridgeItem is updated or added to the local database

![image](https://github.com/user-attachments/assets/c42a773d-2e55-44cf-93f5-3cc03f2eb22d)

This ensures that the user's fridge data is always accessible, even without an internet connection.

## UserViewModel
The `getCurrentUser()` method is enhanced to automatically synchronize fridge data:
- When the app is launched, it updates the local `FridgeItem` database using Firestore if an internet connection is available.
- If the user is offline, fridge items are fetched directly from the local database.
Below is a [schema](https://www.figma.com/design/lWFaH9B1YIkVyy45Cva9yk/M2?node-id=1892-196&t=PfyuwDQnmZDroV8r-1) that illustrates the full process:

![image](https://github.com/user-attachments/assets/21b701c8-4fb0-4942-a27e-d0ee251451a3)


## Sign-In Process
The `getCurrentUser()` method is called only once after the end of the sign-in animation, ensuring a seamless user experience without redundant updates.

---

# Why was this change made?

## Fridge Item Local Storage
This change ensures that users have uninterrupted access to their fridge's contents regardless of their internet connectivity. It enhances user experience by providing a fallback mechanism to access critical data offline.

## UserViewModel
The update simplifies data synchronization and improves efficiency by ensuring the latest fridge data is always available. It also streamlines the logic for offline and online data fetching, reducing the complexity of handling missing data.

## Sign-In Process
Calling the `getCurrentUser()` method after the animation ensures that user data synchronization is not intrusive and aligns with the app's user flow, providing a smoother experience.

---

# Note

When the user is offline and makes changes to their fridge, all API calls to Firestore are cached locally. These changes are automatically sent to Firestore the next time the user has an internet connection. For example, if the user edits the quantity of a fridge item while offline, the updates will be synchronized seamlessly once the connection is restored.

For more details, you can refer to the [official Firebase documentation on enabling offline capabilities](https://firebase.google.com/docs/firestore/manage-data/enable-offline).

---
# Checklist
- [x] Tests added.
- [x] Documentation updated.
- [x] All CI checks passed.
- [x] Linked issues/features (if applicable): #262, #264

AdriB02 commented on app/src/main/java/com/android/sample/ui/authentication/SignInScreen.kt:
```diff
@@ -188,7 +190,10 @@ fun SignInScreen(navigationActions: NavigationActions, userViewModel: UserViewMo
           // Show the loading animation if the user is registered
           if (Firebase.auth.currentUser != null || registered.value) {
             LoadingAnimation(
-                onFinish = { navigationActions.navigateTo(Screen.SWIPE) },
+                onFinish = {
+                  userViewModel.getCurrentUser(isConnected)
+                  navigationActions.navigateTo(Screen.SWIPE)
+                },
```
This part hasn’t been tested yet. It might be a good idea to try returning a mocked user to satisfy the condition and properly test this part of the code. You can refer to this post for guidance on mocking the user: https://stackoverflow.com/questions/61408137/mock-implementation-of-firebase-auth-methods-after-mocking-the-firebase-module.

What do you think?

Ylanv replied:
I’ll wait to check with @CapraSnitzel since the issue is assigned to him. He might be the one handling the testing. Let me know if there’s anything else I can assist with in the meantime!

AdriB02 commented on app/src/test/java/com/android/sample/model/ingredient/local/RoomIngredientRepositoryTest.kt:
You’ve missed testing whether line 70 
```kotlin
val ingredient = ingredientDAO.get(barCode).toIngredient()
```
throws an exception. I think it would be a good idea to test this scenario. What do you think?

Ylanv replied:
Done! I tested the cases where `.get(...)` throws an error and where `toIngredient()` throws an error.

AdriB02 replied:
Nice !

AdriB02 commented on app/src/main/java/com/android/sample/ui/authentication/SignInScreen.kt:
```diff
@@ -188,7 +190,10 @@ fun SignInScreen(navigationActions: NavigationActions, userViewModel: UserViewMo
             LoadingAnimation(
-                onFinish = { navigationActions.navigateTo(Screen.SWIPE) },
+                onFinish = {
+                  userViewModel.getCurrentUser(isConnected)
```
Nice catch ! I think this also fix the issue #270. Is it possible to notify @CapraSnitzel ?

Ylanv replied:
Yes, I just want to discuss it further with him, as there are some points I need clarification on. For example, why we need to call `getCurrentUser` each time we navigate to the user account screen. I’ll bring this up during the next meeting tomorrow and get back to you afterward. Thank you for your patience!

AdriB02 reviewed (changes requested):
Hi @Ylanv,

First of all, great job on your PR! The code is well-written, clear, and easy to follow. The feature you added is a valuable improvement to the app, and I'm really proud of the effort you put into implementing it.

I did encounter two issues during testing that I wanted to bring to your attention:

1. App Crash with Certain Ingredients:
When testing the app with the ingredient "Eau de source", the app crashes immediately after login. This could be caused by an uninitialized value or a missing field in the local database related to this ingredient.

2. Offline Fridge State Issue with "Kròlewska mieszanka orzechów":
The ingredient "Kròlewska mieszanka orzechów" appears in the online fridge but doesn't persist when the app is launched in offline mode. I suspect this might be due to incorrect saving to the database when the ingredient is added. It's possible these two bugs are linked, stemming from how ingredient data is being saved.

#### Suggestion
I have a small suggestion that could improve the user experience:

- #### Add an "Offline Mode" Button on the Sign-In Screen:
   Right now, if a user was not previously connected in online mode, they cannot access the swipe page because the login will fail indefinitely. Would it be possible to include a button on the sign-in screen to allow users to enter the app in offline mode? This would enhance usability for users with unstable connections or situations where login fails.

Thank you again for your excellent work on this PR! Your attention to detail and thoughtful coding are much appreciated. Keep up the great work!

Looking forward to your thoughts.

Ylanv commented on app/src/main/java/com/android/sample/model/user/UserViewModel.kt:
```diff
@@ -115,21 +119,57 @@ class UserViewModel(
+                    if (ingredient != null) {
+                      updateList(_fridgeItems, Pair(fridgeItem, ingredient), add = true)
+                    } else {
+                      Log.e(LOG_TAG, NOT_FOUND_INGREDIENT_IN_DATABASE_ERROR)
```
Yes!

Ylanv commented on app/src/main/java/com/android/sample/model/ingredient/IngredientViewModel.kt:
```diff
@@ -164,6 +164,7 @@ class IngredientViewModel(
         dispatcher,
         onSuccess = {
           repository.addDownload(ingredient)
+          Log.d(INGREDIENT_VIEWMODEL_LOG_TAG, "Ingredient downloaded successfully : $ingredient")
```
I decided to keep the log statements for now until the rest of the implementation is approved. Removing them each time and then rewriting them if a bug is discovered can be time-consuming. Once everything is finalized and clean, I will remove all the `Log.d` statements to ensure a tidy codebase.  

Thank you for your understanding!

Ylanv commented on app/src/main/java/com/android/sample/model/ingredient/EntityMapper.kt:
```diff
@@ -28,6 +30,7 @@ fun Ingredient.toEntity(): IngredientEntity {
  */
 fun IngredientEntity.toIngredient(): Ingredient {
   val gson = Gson()
+  Log.d("EntityMapper", "toIngredient : $this")
```
[See](https://github.com/PlateSwipe/PlateSwipe/pull/278#discussion_r1875637933).

Ylanv commented on app/src/main/java/com/android/sample/model/ingredient/EntityMapper.kt:
```diff
@@ -11,6 +12,7 @@ import com.google.gson.Gson
  */
 fun Ingredient.toEntity(): IngredientEntity {
   val gson = Gson()
+  Log.d("EntityMapper", "toEntity : $this")
```
[See](https://github.com/PlateSwipe/PlateSwipe/pull/278#discussion_r1875637933).

Ylanv commented on app/src/main/java/com/android/sample/model/fridge/localData/FridgeItemDao.kt:
```diff
@@ -0,0 +1,18 @@
+
+  @Delete suspend fun delete(fridgeItem: FridgeItemEntity)
+
+  @Query("SELECT * FROM $FRIDGE_DB_COL_NAME_TABLE") suspend fun getAll(): List<FridgeItemEntity>
```
Yes, exactly! Please note that the `@Insert(onConflict = OnConflictStrategy.REPLACE)` annotation for the `suspend fun insert(fridgeItem: FridgeItemEntity)` method replaces an item in case of a conflict. This approach is also utilized as a method to update the database.

AdriB02 replied:
Thanks for clarifying!

Ylanv commented on app/src/main/java/com/android/sample/model/fridge/localData/FridgeItemEntity.kt:
```diff
@@ -0,0 +1,23 @@
+import com.android.sample.resources.C.Tag.FRIDGE_DB_COL_NAME_TABLE
+import java.time.LocalDate
+
+@Entity(
+    tableName = FRIDGE_DB_COL_NAME_TABLE,
+    primaryKeys = [FRIDGE_DB_COL_NAME_ID, FRIDGE_DB_COL_NAME_EXPIRATION_DATE])
+data class FridgeItemEntity(
+    @ColumnInfo(name = FRIDGE_DB_COL_NAME_ID) val id: String,
+    @ColumnInfo(name = FRIDGE_DB_COL_NAME_QTY) val quantity: Int,
+    @ColumnInfo(name = FRIDGE_DB_COL_NAME_EXPIRATION_DATE) val expirationDate: LocalDate
+)
```
Yes exactly, as explained in the official [documentation](https://developer.android.com/training/data-storage/room/defining-data)

>  to store your app's data, you define entities to represent the objects that you want to store. Each entity corresponds to a table in the associated Room database, and each instance of an entity represents a row of data in the corresponding table.

Ylanv commented on app/src/main/java/com/android/sample/model/fridge/localData/RoomFridgeItemRepository.kt:
```diff
@@ -0,0 +1,31 @@
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+class RoomFridgeItemRepository(
+    private val fridgeItemDao: FridgeItemDao,
+    private val dispatcher: CoroutineDispatcher
+) : FridgeItemLocalRepository {
+
+  override fun add(fridgeItem: FridgeItem) {
+    CoroutineScope(dispatcher).launch { fridgeItemDao.insert(fridgeItem.toFridgeItemEntity()) }
+  }
+
+  override fun delete(fridgeItem: FridgeItem) {
+    CoroutineScope(dispatcher).launch { fridgeItemDao.delete(fridgeItem.toFridgeItemEntity()) }
+  }
+
+  override fun getAll(onSuccess: (List<FridgeItem>) -> Unit, onFailure: (Exception) -> Unit) {
+    CoroutineScope(dispatcher).launch {
+      try {
+        val fridgeItems = fridgeItemDao.getAll().map { it.toFridgeItem() }
+        onSuccess(fridgeItems)
+      } catch (e: Exception) {
+        onFailure(e)
+      }
```
Honestly, I'm passing the dispatcher as an argument because SonarCloud flagged it as an issue otherwise, citing maintainability concerns. To address this, I'm passing `Dispatchers.IO` as a parameter, as it is appropriate for writing to the device's storage.

Ylanv commented on app/src/main/java/com/android/sample/utils/NetworkUtils.kt:
```diff
@@ -0,0 +1,23 @@
+   * @param context The context used to get the ConnectivityManager.
+   * @return True if the network is available, false otherwise.
+   */
+  fun isNetworkAvailable(context: Context): Boolean {
+    val connectivityManager =
+        context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+    val network = connectivityManager.activeNetwork ?: return false
+    val networkCapabilities = connectivityManager.getNetworkCapabilities(network) ?: return false
+    return networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+  }
```
That's a great question! Since I currently rely on 4G and don’t have Wi-Fi at home, I didn’t test it under those specific conditions, to be honest. However, I did test it on the emulator with Wi-Fi and Airplane mode enabled, and the result was that `isNetworkAvailable` returned `true`.

AdriB02 replied:
Yes, I tested it as well, and it worked in the same way. I think this is an effective method to implement it!

Ylanv commented:
> 1. App Crash with Certain Ingredients:
>    When testing the app with the ingredient "Eau de source", the app crashes immediately after login. This could be caused by an uninitialized value or a missing field in the local database related to this ingredient.
> 2. Offline Fridge State Issue with "Kròlewska mieszanka orzechów":
>    The ingredient "Kròlewska mieszanka orzechów" appears in the online fridge but doesn't persist when the app is launched in offline mode. I suspect this might be due to incorrect saving to the database when the ingredient is added. It's possible these two bugs are linked, stemming from how ingredient data is being saved.


First of all, thank you for your review and detailed feedback!  

### **Problem**  
For the first and second issue, I believe the problem stemmed from adding ingredients via the search bar. When implementing this feature, I assumed that searching for an ingredient by name followed the same process as scanning an ingredient. This assumption led me to believe that the ingredient would be downloaded into our Firestore collection, but that is only the case for the scanning feature.  

As a result, when searching for an ingredient by name, the `uid` field of the `Ingredient` object is `null` (since it does not exist in a Firestore collection). This caused the ingredient to not be downloaded locally, as I was using the `uid` as the identifier for the local database. When going offline, the app attempted to retrieve an item locally that had not been downloaded, leading to the crash.  

### **Solution**  
After discussing with @RagdollGoDown, we decided not to change how the search-by-name feature works. Instead, I implemented a solution where a `uid` is derived from the barcode if the ingredient does not already have a `uid` (indicating it comes from OpenFoodFacts). This approach seems to resolve the issue.  

### **Request**  
I would be very grateful if you could test again to confirm whether the issue is now fixed on your side. On my side, this change appears to have resolved the problem.  

Thank you again for your time and the detailed report, it’s greatly appreciated!

Ylanv commented:
>  Suggestion
> I have a small suggestion that could improve the user experience:
> 
> Add an "Offline Mode" Button on the Sign-In Screen:
> Right now, if a user was not previously connected in online mode, they cannot access the swipe page because the login will fail indefinitely. Would it be possible to include a button on the sign-in screen to allow users to enter the app in offline mode? This would enhance usability for users with unstable connections or situations where login fails.

Concretely, if the user tries to connect for the first time without an internet connection, it will not work. This is because we need an active internet connection to fetch Google account information from Firebase, and therefore, this step cannot be skipped. Allowing the app to be used without an account is not intended, as all features are directly linked to the user’s account.  

That said, it’s important to note that a user can log in without an internet connection if their account has already been cached by Firebase, which happens automatically.

I agree that if we were using custom identification, it might be possible to cache the entire process (e.g., creating an account offline, interacting with the app, etc.) and then synchronize it with the server once a connection is available. However, implementing such a feature would undoubtedly require significant time and effort. Given that we are approaching the deadline, I believe it would be unwise to attempt such a complex and potentially risky change at this stage.

AdriB02 commented on app/src/main/java/com/android/sample/model/user/UserViewModel.kt:
```diff
@@ -115,21 +119,57 @@ class UserViewModel(
+   * @param isConnected Boolean indicating if the device is connected to the internet.
+   * @param user The user whose fridge items are to be handled.
+   */
+  fun handleFridgeItem(isConnected: Boolean, user: User) {
```
That works for me. Making it internal seems like a logical decision given its intended usage—thanks for clarifying! By the way, testing private methods is definitely possible; @CapraSnitzel has done it. We can discuss this further in tomorrow's meeting if you'd like.

AdriB02 reviewed (changes requested):
Hi @Ylanv,
great job fixing these issues! I tested it multiple times, and the bugs no longer reappear. Well done! 
However, I noticed something odd: the ingredient images take a long time to load. Do you know why this might be happening? On the main branch, the images load instantly, but on your branch, they take 10-15 seconds to appear. It might be a good idea to investigate and address this before merging with the main branch.

Ylanv commented:
> Hi @Ylanv, great job fixing these issues! I tested it multiple times, and the bugs no longer reappear. Well done! However, I noticed something odd: the ingredient images take a long time to load. Do you know why this might be happening? On the main branch, the images load instantly, but on your branch, they take 10-15 seconds to appear. It might be a good idea to investigate and address this before merging with the main branch.

Hello,  

I tested it on my phone, and I didn’t notice any issues with the time it takes for the image to load.  

[Link to image or video demonstrating your test](https://github.com/user-attachments/assets/030c373d-cfe4-4f30-8cd8-c6616f481029)

Could you please provide more detailed information about the steps you took to achieve the result you encountered? This would help me replicate the behavior and investigate further.  

Thank you!

AdriB02 reviewed (changes requested):
Hi @Ylanv,

That’s a bit unusual, but it’s working correctly again for now, perhaps the database was overwhelmed at the time.

I did encounter two new bug by testing the ingredient images:
1. First, when I added an ingredient to the fridge, then quickly exited the app and reopened it in offline mode. As shown in the attached video, the app crashes in offline mode but works again when restarted in online mode.


https://github.com/user-attachments/assets/cbe4791e-47d7-4f5e-9f22-f9bb694398ff


It seems like the crash happens because the fridge ingredient wasn’t properly saved or synced before I closed the app. It might be a good idea to prevent crashes in such cases by showing an empty fridge instead if the ingredients aren’t found. What do you think?

Here is the error: 
```code
E  FATAL EXCEPTION: DefaultDispatcher-worker-2
Process: com.android.sample, PID: 7080
 java.lang.NullPointerException: Parameter specified as non-null is null: method com.android.sample.model.ingredient.EntityMapperKt.toIngredient, parameter <this>
at com.android.sample.model.ingredient.EntityMapperKt.toIngredient(Unknown Source:8)
at com.android.sample.model.ingredient.localData.RoomIngredientRepository$getByBarcode$1.invokeSuspend(RoomIngredientRepository.kt:70)
at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)
at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)
at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@31d10d7, Dispatchers.IO]
```

2. Sometimes, when I start the app quickly and navigate to the fridge, it displays an empty fridge. It seems like the fridge isn't initialized properly because refreshing (by going to the account screen and returning to the fridge) makes the correct fridge appear. This might be related to getCurrentUser, perhaps the function responsible for initializing the fridge isn’t being called correctly. The issue doesn’t always happen, so it’s challenging to debug and provide more concrete insights into the root cause.

Here’s a video demonstrating how the fridge refreshes after initially displaying as empty:

https://github.com/user-attachments/assets/de123972-8e8f-42ad-9d3d-98519056c3a1


I have another suggestion regarding certain functionalities in offline mode. I noticed that you’ve already disabled the swipe page in offline mode, which is great. Perhaps it would also be a good idea to disable the ability to create a recipe while offline. What do you think?

Thanks for your work on this! I hope these bugs won’t be too difficult to resolve!

Ylanv commented:
Thank you for your detailed bug report, it was incredibly helpful in uncovering some hidden issues I hadn’t considered.  

For the first problem, as you correctly pointed out, the app crashes because it doesn't have enough time to save locally before attempting to fetch the data. While I may not be able to completely resolve the root cause, I will ensure the error is handled gracefully to prevent the app from crashing.  

Regarding the second issue, I haven’t been able to reproduce the bug on my end. However, here’s what should happen: after the animation completes, `getCurrentUser()` is called. If you have a large number of ingredients in the fridge, it might take around a second to fetch them all. During this brief period, the fridge may appear empty, but it should update as soon as all ingredients are retrieved from the local database.  

Thank you again for bringing these issues to my attention! Let me know if you experience anything further.

Ylanv commented:
> I have another suggestion regarding certain functionalities in offline mode. I noticed that you’ve already disabled the swipe page in offline mode, which is great. Perhaps it would also be a good idea to disable the ability to create a recipe while offline. What do you think?

Yes, I can take care of it. I just need to confirm with the Product Owner during tomorrow's meeting.

AdriB02 commented:
Hi @Ylanv,
The second issue no longer appears after I pulled your latest commit. It seems that the changes you made resolved it as well, great job!

However, I encountered a new issue: the fridge updates automatically, as if the ingredients were updated by themselves without any action. Here’s a video demonstrating the bug::

https://github.com/user-attachments/assets/f61b7b3f-dae1-4c95-8301-f52a0ff2f7f7

I suspect this issue might be related to improper storage of ingredients in the local database. It only occurs with ingredients searched by name and not those scanned by barcode. This behavior makes me think it could be tied to incorrect data provided by OpenFoodFacts.

Let me know your thoughts!

Ylanv commented:
**@AdriB02**, I waited for your updated implementation of the Fridge before addressing the issues. Now, the latest version aligns with your new implementation. I have also updated the PR description to include a revised schema illustrating how the local database for `FridgeItem` is updated.

Additionally, I tested all the potential bugs you identified with the new implementation and did not encounter any issues. Please let me know if there’s anything else you'd like me to check!

Ylanv commented:
@AdriB02, I waited for your updated implementation of the Fridge before addressing the issues. Now, the latest version aligns with your new implementation. I have also updated the PR description to include a revised schema illustrating how the local database for FridgeItem is updated.

Additionally, I tested all the potential bugs you identified with the new implementation and did not encounter any issues. Please let me know if there’s anything else you'd like me to check!

AdriB02 commented:
Hi @Ylanv,

First, thank you so much for fixing these bugs! We really appreciate the effort you've put into this. It's great to see that the app no longer crashes, that's a significant improvement!

However, I encountered a new issue during testing. When I update an ingredient and save the changes (both in online and offline modes), then restart the app in offline mode, the ingredient disappears.

- **Demo of the bug:**

https://github.com/user-attachments/assets/18320028-61ca-45aa-8b4f-6b4e2bc0ea1c

- **Here is the barcode of the ingredient I added in the fridge:**

<img src="https://github.com/user-attachments/assets/381d3075-4743-486c-8f3a-58fc69555915" width="200" />

This seems to be related to a bug that occurs when editing in both (offline and online) modes and restarting in offline mode. It appears to be caused by the updated element not being added to the local database. The issue might be in the `updateIngredientFromFridge` method, particularly with how _currentEditingFridgeIngredient (the ingredient before modification) and the new Ingredient parameter are handled.

Let me know if you need more details or testing on this!

Ylanv commented:
Thank you again for your detailed testing! I sincerely apologize for the extra time you've spent debugging due to this issue. Your effort and patience mean a lot. I’m confident this time the fix is solid, and I’d like to explain what caused the problem and how it’s been addressed.  

Previously, when handling the update of a `FridgeItem`, I used the following logic:  

```kotlin
val existingItem = fridgeItemDao.getByIdAndExpirationDate(id, newExpirationDate)
```  

If an item with the new parameters already existed, I was updating this existing item by summing the quantities and then deleting the original item since it became redundant (as the new item represented the combination of the two). I designed this approach assuming the item being updated would always result in a distinct new item.  

However, I didn’t account for the scenario where the `newExpirationDate` is the same as the current one. In such cases, the query:  

```kotlin
val existingItem = fridgeItemDao.getByIdAndExpirationDate(id, newExpirationDate)
```  

would fetch the very item I intended to update. The subsequent logic would then update the item and immediately delete it, causing the bug.  

To fix this, I’ve added a special condition to handle cases where the current expiration date matches the new one:  

```kotlin
if (currentExpirationDate == newExpirationDate) {
    fridgeItemDao.update(FridgeItemEntity(id, newQuantity, newExpirationDate))
    return@launch
}
```  

This ensures the item is updated correctly without triggering the redundant delete logic.  

Thank you again for your thorough testing and understanding.

AdriB02 reviewed (approved):
Thank you so much for fixing the last bug I reported!

After extensive testing, I couldn’t find any issues in the fridge functionality. I believe it’s now the right time to approve this new feature and merge it into the main branch.

Thank you again for adding this essential feature to the app. The project is shaping up really well, and I’m genuinely proud of what we’ve accomplished together. I’ll go ahead and merge this into the main branch now!
